### PR TITLE
change table column to 4096 with optimization

### PR DESCRIPTION
**Pull request by XIAO-HOU**
### What changes were proposed in this pull request?
Modify the maximum number of columns in a single table to 4096.

### Why are the changes needed?
Solve #910 .

### Will break the compatibility? How if so?
You should init ObRowWriter before using it and reset it when you needed.

### Does this PR introduce any user-facing change?
No.

### Checklist

- [x] I've run the tests to see all new and existing tests pass.
- [x] If this Pull Request resolves an issue, I linked to the issue in the text above.
- [ ] I've informed the technical writer about the documentation change if necessary.

### CODE REVIEW

File: src/sql/resolver/dml/ob_insert_resolver.h [hunk 1]
```diff
@@ -119,7 +119,7 @@ class ObInsertResolver : public ObDMLResolver {
   {
     is_oracle_tmp_table_ = is_temp_table;
   }
-  common::hash::ObPlacementHashSet<uint64_t>& get_insert_column_ids()
+  common::hash::ObPlacementHashSet<uint64_t, 8209>& get_insert_column_ids()
```
Comment (haitaoy) [inline src/sql/resolver/dml/ob_insert_resolver.h:122]: add comment tells why is 8209

File: src/sql/resolver/dml/ob_insert_resolver.h [hunk 2]
```diff
@@ -132,7 +132,7 @@ class ObInsertResolver : public ObDMLResolver {
   DISALLOW_COPY_AND_ASSIGN(ObInsertResolver);
 
 private:
-  common::hash::ObPlacementHashSet<uint64_t> insert_column_ids_;
+  common::hash::ObPlacementHashSet<uint64_t, 8209> insert_column_ids_;
```
Comment (haitaoy) [inline src/sql/resolver/dml/ob_insert_resolver.h:135]: same with prev

File: src/storage/blocksstable/ob_lob_data_writer.cpp [hunk 3]
```diff
@@ -304,7 +304,7 @@ int ObLobMacroBlockWriter::init_header(
           column_types_[i] = column_types.at(i);
           column_checksum_[i] = 0;
           if (i < header_->rowkey_column_count_) {
-            column_orders_[i] = desc_.column_orders_[i];
+            column_orders_[i] = desc_.column_orders_.get_buf()[i];
```
Comment (haitaoy) [inline src/storage/blocksstable/ob_lob_data_writer.cpp:307]: can put get_buf inside ObRowBuffer<T>?

File: src/storage/blocksstable/ob_lob_merge_writer.cpp [hunk 4]
```diff
@@ -410,10 +413,15 @@ int ObLobMergeWriter::write_lob_obj(
...
-  result_row_.row_val_.count_ = row.row_val_.count_;
-  MEMCPY(buffer_, row.row_val_.cells_, sizeof(ObObj) * row.row_val_.count_);
+  int64_t request_count = data_store_desc_->row_column_count_;
+  if OB_FAIL(buffer_.reserve(request_count)) {
+    STORAGE_LOG(WARN, "fail to reserve memory for buffer_, ", K(ret));
+  } else {
+    result_row_ = row;
+    result_row_.row_val_.cells_ = buffer_.get_buf();
+    result_row_.row_val_.count_ = row.row_val_.count_;
+    MEMCPY(buffer_.get_buf(), row.row_val_.cells_, sizeof(ObObj) * row.row_val_.count_);
```
Comment (haitaoy) [inline src/storage/blocksstable/ob_lob_merge_writer.cpp:423]: add check ensure buffer is enough

File: src/storage/blocksstable/ob_macro_block.cpp [hunk 5]
```diff
@@ -199,14 +206,22 @@ int ObDataStoreDesc::init(const ObTableSchema& table_schema, const int64_t data_
...
-            column_orders_[i] = column_list.at(i).col_order_;
+        if (OB_FAIL(column_ids_.reserve(row_column_count_))) {
+          STORAGE_LOG(WARN, "fail to reserve memory for column_ids_, ", K(ret));
+        } else if (OB_FAIL(column_types_.reserve(row_column_count_))) {
+          STORAGE_LOG(WARN, "fail to reserve memory for column_types_, ", K(ret));
+        } else if (OB_FAIL(column_orders_.reserve(row_column_count_))) {
+          STORAGE_LOG(WARN, "fail to reserve memory for column_orders_, ", K(ret));
+        } else {
+          for (int64_t i = 0; OB_SUCC(ret) && i < column_list.count(); ++i) {
+            if (i >= common::OB_MAX_COLUMN_NUMBER) {
```
Comment (haitaoy) [inline src/storage/blocksstable/ob_macro_block.cpp:217]: check column_list.. is enough

File: src/storage/blocksstable/ob_macro_block.cpp [hunk 6]
```diff
@@ -265,6 +280,7 @@ void ObDataStoreDesc::reset()
   major_working_cluster_version_ = 0;
   iter_complement_ = false;
   is_unique_index_ = false;
+  allocator_.reset();
```
Comment (haitaoy) [inline src/storage/blocksstable/ob_macro_block.cpp:283]: it's risky here,  prev codes just reset some state and not release memory from allocator, should check whether some pointers are still used after reset

File: src/sql/resolver/dml/ob_insert_resolver.h (hunk 1 above)
Comment (XIAO-HOU) [reply src/sql/resolver/dml/ob_insert_resolver.h:122]: In the original implementation, the default length of ObPlacementHashSet is 1039, which is the first prime number greater than 512. After changing the max column to 4096, I think changing it to 8209 (the first prime number greater than 4096) is a good choice.

File: src/sql/resolver/dml/ob_insert_resolver.h (hunk 2 above)
Comment (XIAO-HOU) [reply src/sql/resolver/dml/ob_insert_resolver.h:135]: fixed.

File: src/storage/blocksstable/ob_macro_block.cpp (hunk 6 above)
Comment (XIAO-HOU) [reply src/storage/blocksstable/ob_macro_block.cpp:283]: In the original code, I find that the allocator_ is not used in ObDataStoreDesc functions. But now I use it to allocate memory for column_ids_... So I think it should be reset in the reset() function.

File: src/storage/blocksstable/ob_block_index_intermediate.cpp [hunk 7]
```diff
@@ -80,14 +81,20 @@ int ObBlockIntermediateBuilder::init(const int64_t rowkey_column_count)
...
-    intermediate_row_.row_val_.cells_ = new (obj_buf_) ObObj[rowkey_column_count_ + 1];
-    intermediate_row_.row_val_.count_ = rowkey_column_count_ + 1;
-    intermediate_row_.row_val_.projector_ = nullptr;
-    intermediate_row_.row_val_.projector_size_ = 0;
-    is_inited_ = true;
+    if (OB_FAIL(obj_buf_.reserve(rowkey_column_count_ + 1))) {
+      STORAGE_LOG(WARN, "fail to reserve memory for obj_buf_, ", K(ret));
+    } else {
+      intermediate_row_.flag_ = common::ObActionFlag::OP_ROW_EXIST;
+      intermediate_row_.row_val_.cells_ = new (obj_buf_.get_buf()) ObObj[rowkey_column_count_ + 1];
```
[the remark this answers is not in the document]
Comment (XIAO-HOU) [reply src/storage/blocksstable/ob_block_index_intermediate.cpp:92]: fixed.

File: src/storage/blocksstable/ob_lob_data_writer.cpp (hunk 3 above)
Comment (XIAO-HOU) [reply src/storage/blocksstable/ob_lob_data_writer.cpp:307]: Yes, I have changed it.

File: src/storage/blocksstable/ob_lob_merge_writer.cpp [hunk 8]
```diff
@@ -79,6 +80,8 @@ int ObLobMergeWriter::init(const ObMacroDataSeq& macro_start_seq, const ObDataSt
     } else if (!block_write_ctx_.file_handle_.is_valid() &&
                OB_FAIL(block_write_ctx_.file_handle_.assign(data_store_desc.file_handle_))) {
       STORAGE_LOG(WARN, "Failed to assign file handle", K(ret));
+    } else if (OB_FAIL(buffer_.init(&block_write_ctx_.allocator_))) {
```
[the remark this answers is not in the document]
Comment (XIAO-HOU) [reply src/storage/blocksstable/ob_lob_merge_writer.cpp:83]: Ok, I have used a new allocator.

File: src/storage/blocksstable/ob_lob_merge_writer.cpp (hunk 4 above)
Comment (XIAO-HOU) [reply src/storage/blocksstable/ob_lob_merge_writer.cpp:423]: I change the reserved buffer size to row.row_val_.count_ which is always an enough buffer.

File: src/storage/blocksstable/ob_row_writer.h [hunk 9]
```diff
@@ -99,10 +101,12 @@ class ObRowWriter {
...
   common::number::ObNumber tmp_number_;
-  uint16_t column_ids_[common::OB_ROW_MAX_COLUMNS_COUNT];  // for sparse row
-  int8_t column_indexs_8_[common::OB_ROW_MAX_COLUMNS_COUNT];
-  int16_t column_indexs_16_[common::OB_ROW_MAX_COLUMNS_COUNT];
-  int32_t column_indexs_32_[common::OB_ROW_MAX_COLUMNS_COUNT];
+  storage::ObRowBuffer<uint16_t> column_ids_;
+  storage::ObRowBuffer<int8_t> column_indexs_8_;
+  storage::ObRowBuffer<int16_t> column_indexs_16_;
+  storage::ObRowBuffer<int32_t> column_indexs_32_;
+  common::ObArenaAllocator allocator_;
```
[the remark this answers is not in the document]
Comment (XIAO-HOU) [reply src/storage/blocksstable/ob_row_writer.h:108]: done.

File: src/storage/ob_i_store.h [hunk 10]
```diff
@@ -1573,6 +1573,83 @@ struct ObRowStat {
...
+      capacity_ = LOCAL_BUFFER_SIZE;
+      is_inited_ = true;
+    }
+    return ret;
+  }
+
+  void reset()
+  {
+    if (OB_NOT_NULL(data_) && reinterpret_cast<T*>(data_buf_) != data_) {
+      allocator_->free(data_);
```
[the remark this answers is not in the document]
Comment (XIAO-HOU) [reply src/storage/ob_i_store.h:1612]: added.

File: src/storage/ob_i_store.h [hunk 11]
```diff
@@ -1573,6 +1573,83 @@ struct ObRowStat {
...
+    }
+    data_ = NULL;
+    allocator_ = NULL;
+    capacity_ = 0;
+    is_inited_ = false;
+  }
+
+  inline int reserve(int capacity) {
+    int ret = OB_SUCCESS;
```
[the remark this answers is not in the document]
Comment (XIAO-HOU) [reply src/storage/ob_i_store.h:1621]: done.

File: src/storage/ob_i_store.h [hunk 12]
```diff
@@ -1573,6 +1573,83 @@ struct ObRowStat {
...
+      int64_t new_size = capacity * sizeof(T);
+      T* new_data = reinterpret_cast<T*>(allocator_->alloc(new_size));
+      if (OB_ISNULL(new_data)) {
+        ret = OB_ALLOCATE_MEMORY_FAILED;
+        STORAGE_LOG(WARN, "failed to alloc new_data", K(ret));
+      } else {
+        if (reinterpret_cast<T*>(data_buf_) != data_) {
+          allocator_->free(data_);
+        }
+        data_ = new_data;
```
[the remark this answers is not in the document]
Comment (XIAO-HOU) [reply src/storage/ob_i_store.h:1632]: done.

File: src/storage/ob_i_store.h [hunk 13]
```diff
@@ -1573,6 +1573,83 @@ struct ObRowStat {
...
+  }
+
+  inline T* get_buf() const
+  {
+    return data_;
+  }
+
+
+private:
+  char data_buf_[LOCAL_BUFFER_SIZE * sizeof(T)];
```
[the remark this answers is not in the document]
Comment (XIAO-HOU) [reply src/storage/ob_i_store.h:1646]: changed.

File: src/storage/ob_i_store.h [hunk 14]
```diff
@@ -1573,6 +1573,83 @@ struct ObRowStat {
   TO_STRING_KV(K_(base_row_count), K_(inc_row_count), K_(merge_row_count), K_(result_row_count), K_(filt_del_count));
 };
 
+template <typename T, int64_t LOCAL_BUFFER_SIZE=64>
+class ObRowBuffer {
```
Comment (haitaoy) [inline src/storage/ob_i_store.h:1577]: ObRowBuffer is not a good name, may be ObDynamicBuffer or..?

File: src/storage/blocksstable/ob_macro_block.cpp (hunk 5 above)
Comment (XIAO-HOU) [reply src/storage/blocksstable/ob_macro_block.cpp]: changed to "i >= row_column_count_"

File: src/storage/ob_i_store.h (hunk 14 above)
Comment (XIAO-HOU) [reply src/storage/ob_i_store.h]: changed to ObDynamicBuffer

File: src/storage/ob_i_store.h [hunk 15]
```diff
@@ -1573,6 +1573,101 @@ struct ObRowStat {
...
+
+  virtual ~ObDynamicBuffer()
+  {
+    reset();
+  }
+  
+  int init(ObIAllocator *allocator)
+  {
+    int ret = OB_SUCCESS;
+    if (OB_UNLIKELY(is_inited_)) {
```
Comment (haitaoy) [inline src/storage/ob_i_store.h:1594]: there is IS_INIT macro

File: src/storage/ob_i_store.h [hunk 16]
```diff
@@ -1573,6 +1573,101 @@ struct ObRowStat {
...
+      ret = OB_INIT_TWICE;
+      STORAGE_LOG(WARN, "init twice", K(ret));
+    } else if (OB_ISNULL(allocator)) {
```
Comment (haitaoy) [inline src/storage/ob_i_store.h:1597]: there is OB_ISNULL/OB_NOT_NULL macro

File: src/storage/blocksstable/ob_micro_block_row_scanner.cpp [hunk 17]
```diff
@@ -469,14 +469,21 @@ int ObMicroBlockRowScanner::init(const ObTableIterParam& param, ObTableAccessCon
                  projector,
                  storage::ObMultiVersionRowkeyHelpper::MVRC_NONE /*multi version col type*/))) {
     STORAGE_LOG(WARN, "Fail to init col map, ", K(ret));
+  } else if (OB_FAIL(obj_buf_.init(context.allocator_))) {
```
Comment (haitaoy) [inline src/storage/blocksstable/ob_micro_block_row_scanner.cpp:472]: context.allocator_'s lifecycle is not enough, in rescan cases, allocator_ will be released in each scan, use stmt_allocator_ instead

File: src/storage/blocksstable/ob_micro_block_row_scanner.cpp [hunk 18]
```diff
@@ -605,6 +615,8 @@ int ObMultiVersionMicroBlockRowScanner::init(
       STORAGE_LOG(WARN, "failed to alloc row", K(ret), K(cell_cnt_));
     } else if (OB_FAIL(nop_pos_.init(*context.allocator_, cell_cnt_))) {
       STORAGE_LOG(WARN, "failed to init nop_pos", K(ret), K(cell_cnt_));
+    } else if (OB_FAIL(tmp_row_obj_buf_.init(context.allocator_))) {
```
Comment (haitaoy) [inline src/storage/blocksstable/ob_micro_block_row_scanner.cpp:618]: same with prev

File: deps/oblib/src/lib/allocator/ob_mod_define.h [hunk 19]
```diff
@@ -485,6 +485,7 @@ LABEL_ITEM_DEF(OB_STORE_FILE_SYSTEM, StoreFileSystem)
 LABEL_ITEM_DEF(OB_HASH_BUCKET_SVR_META_MGR, HashBucSvrMetMg)
 LABEL_ITEM_DEF(OB_HASH_NODE_SVR_META_MGR, HashNodSvrMetMg)
 LABEL_ITEM_DEF(OB_TENANT_FILE_MGR, TenantFileMgr)
+LABEL_ITEM_DEF(OB_LOB_MERGE_WRITER, LobMergeWriter)
```
[the remark this answers is not in the document]
Comment (XIAO-HOU) [reply deps/oblib/src/lib/allocator/ob_mod_define.h]: fixed.

File: src/storage/ob_i_store.h [hunk 20]
```diff
@@ -1573,6 +1573,101 @@ struct ObRowStat {
...
+  }
+
+  inline int reserve(int64_t capacity) {
+    int ret = OB_SUCCESS;
+    if (OB_UNLIKELY(!is_inited_)) {
+      ret = OB_NOT_INIT;
+      STORAGE_LOG(ERROR, "row buffer has not been initialized", K(ret));
+    } else if (capacity < 0) {
+      ret = OB_INVALID_ARGUMENT;
+      STORAGE_LOG(WARN, "invalid arguments", K(ret));
```
[the remark this answers is not in the document]
Comment (XIAO-HOU) [reply src/storage/ob_i_store.h]: fixed.

File: src/storage/blocksstable/ob_macro_block.cpp [hunk 21]
```diff
@@ -289,9 +305,23 @@ int ObDataStoreDesc::assign(const ObDataStoreDesc& desc)
   has_lob_column_ = desc.has_lob_column_;
   is_major_ = desc.is_major_;
   MEMCPY(compressor_name_, desc.compressor_name_, OB_MAX_HEADER_COMPRESSOR_NAME_LENGTH);
-  MEMCPY(column_ids_, desc.column_ids_, sizeof(column_ids_));
-  MEMCPY(column_types_, desc.column_types_, sizeof(column_types_));
-  MEMCPY(column_orders_, desc.column_orders_, sizeof(column_orders_));
+  if (OB_FAIL(column_ids_.init(&allocator_))) {
```
[the remark this answers is not in the document]
Comment (XIAO-HOU) [reply src/storage/blocksstable/ob_macro_block.cpp:311]: fixed.

File: src/storage/blocksstable/ob_macro_block_writer.h [hunk 22]
```diff
@@ -206,10 +207,9 @@ class ObMacroBlockWriter {
   ObColumnMap column_map_;
   ObColumnMap index_column_map_;
   blocksstable::ObMacroBlockReader macro_reader_;
-  char obj_buf_[common::OB_ROW_MAX_COLUMNS_COUNT * sizeof(common::ObObj)];          // for reader to get row
-  char checker_obj_buf_[common::OB_ROW_MAX_COLUMNS_COUNT * sizeof(common::ObObj)];  // for calc or varify checksum, can
-                                                                                    // NOT use same buf of data row
-  ObMicroBlockReader check_flat_reader_;
+  storage::ObDynamicBuffer<ObObj> obj_buf_;          // for reader to get row
+  storage::ObDynamicBuffer<ObObj> checker_obj_buf_;  // for calc or varify checksum, can not use same buf of data row
+  ObMicroBlockReader check_flat_reader_;
```
[the remark this answers is not in the document]
Comment (XIAO-HOU) [reply src/storage/blocksstable/ob_macro_block_writer.h:211]: fixed.

File: src/storage/blocksstable/ob_micro_block_row_getter.cpp [hunk 23]
```diff
@@ -319,31 +335,35 @@ int ObMicroBlockRowGetter::project_cache_row(const ObStoreRowkey& rowkey, const
   const int64_t column_cnt = value.get_column_cnt();
   const ObColumnIndexItem* column_indexs = column_map.get_column_indexs();
   const int64_t request_count = column_map.get_request_count();
-  row_.row_val_.cells_ = reinterpret_cast<ObObj*>(obj_buf_);
-  row_.row_val_.count_ = OB_ROW_MAX_COLUMNS_COUNT;
+  if (OB_FAIL(obj_buf_.reserve(request_count))) {
+    STORAGE_LOG(WARN, "failed to reserve memory for obj_buf_, ", K(ret));
+  } else {
+    row_.row_val_.cells_ = obj_buf_.get_buf();
+    row_.row_val_.count_ = request_count;
```
[the remark this answers is not in the document]
Comment (XIAO-HOU) [reply src/storage/blocksstable/ob_micro_block_row_getter.cpp:342]: added.

File: src/storage/blocksstable/ob_block_index_intermediate.cpp [hunk 24]
```diff
@@ -64,6 +64,7 @@ void ObBlockIntermediateBuilder::reset()
   intermediate_row_.reset();
   header_.reset();
   allocator_.reset();
+  obj_buf_.reset();
```
Comment (haitaoy) [inline src/storage/blocksstable/ob_block_index_intermediate.cpp:67]: must reset obj_buf before allocator

File: src/storage/blocksstable/ob_macro_block.h [hunk 25]
```diff
@@ -53,9 +53,9 @@ struct ObDataStoreDesc {
   bool is_major_;
 
   char compressor_name_[common::OB_MAX_HEADER_COMPRESSOR_NAME_LENGTH];
-  uint64_t column_ids_[common::OB_MAX_COLUMN_NUMBER];
-  common::ObObjMeta column_types_[common::OB_MAX_COLUMN_NUMBER];
-  common::ObOrderType column_orders_[common::OB_MAX_COLUMN_NUMBER];
+  storage::ObDynamicBuffer<uint64_t> column_ids_;
+  storage::ObDynamicBuffer<ObObjMeta> column_types_;
+  storage::ObDynamicBuffer<ObOrderType> column_orders_;
```
Comment (haitaoy) [inline src/storage/blocksstable/ob_macro_block.h:58]: in deconstructor allocator_ will be released before buffer?

Comment (XIAO-HOU) [reply src/storage/blocksstable/ob_macro_block.h:59]: fixed.

**Review by chaser-ch — APPROVED** [review]

**Review by haitaoy — APPROVED** [review]
LGTM

**Comment by hnwyllmm** [comment]
merged

**Pull request merged**